python/mooseutils/PerfGraphReporterReader.py

```diff
@@ -0,0 +1,312 @@
+#* This file is part of the MOOSE framework
+#* https://www.mooseframework.org
+#*
+#* All rights reserved, see COPYRIGHT for full restrictions
+#* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+#*
+#* Licensed under LGPL 2.1, please see LICENSE for details
+#* https://www.gnu.org/licenses/lgpl-2.1.html
+
+from mooseutils.ReporterReader import ReporterReader
+
+class PerfGraphNode:
+    """
+    A node in the graph for the PerfGraphReporterReader.
+    These should really only be constructed internally within
+    the PerfGraphReporterReader.
+
+    Inputs:
+        id[int]: The ID in the JSON graph
+        parents[PerfGraphNode]: The node parent (None if root)
+        data[dict]: The object that contains the full graph
+    """
+    def __init__(self, id, parent, data):
+        node_data = None
+        for entry in data:
+            if 'id' not in entry or not isinstance(entry['id'], int):
+                raise Exception('Entry missing ID\n{}'.format(entry))
+            if entry['id'] == id:
+                if node_data:
+                    raise Exception('Duplicate ID {} found'.format(id))
+                node_data = entry
+        if node_data is None:
+            raise Exception('Failed to find node with ID {}'.format(id))
+
+        # Validate that the data in the node is as we expect
+        self._validateNodeData(node_data)
+
+        # Sets self._id, self._level, self._memory, etc...
+        for key, val in node_data.items():
+            if key not in ['parent_id', 'children_ids']:
+                setattr(self, '_' + key, val)
+
+        if parent is not None and not isinstance(parent, PerfGraphNode):
+            raise TypeError('parent is not of type "PerfGraphNode"')
+        self._parent = parent
+
+        # Recursively add all of the children
+        self._children = []
+        for child_id in node_data.get('children_ids', []):
+            self._children.append(PerfGraphNode(child_id, self, data))
```

Comment:
And now you're trying to recreate the tree?

Reply:
Commenting above